src/components/forms/CharacterCount/CharacterCount.tsx

```diff
@@ -0,0 +1,165 @@
+  const remainingSingularFormat = '$0 character left'
+  const overSingularFormat = '$0 character over limit'
+  const overPluralFormat = '$0 characters over limit'
+
+  const limitMessage = (len: number): string => {
+    let limitMessage
+    if (len === 0) {
+      return emptyMessageFormat
+    } else if (len <= maxNum) {
+      switch (maxNum - len) {
+        case 1:
+          return remainingSingularFormat.replace('$0', '1')
+        default:
+          return remainingPluralFormat.replace('$0', (maxNum - len).toString())
+      }
+    } else {
+      switch (len - maxNum) {
+        case 1:
+          return overSingularFormat.replace('$0', '1')
+        default:
+          return overPluralFormat.replace('$0', (len - maxNum).toString())
+      }
+    }
+  }
+
+  const [state, setState] = useState({
+    characterLength: defaultLength,
+    limitMessage: limitMessage(defaultLength),
+    invalidMessage: defaultLength > maxNum,
+  })
+
+  const classes = classnames('usa-character-count__field', className)
+  const messageClasses = classnames(
+    'usa-hint',
+    'usa-character-count__message',
+    { 'usa-character-count__message--invalid': state.invalidMessage }
+  )
+
+  const messageId = id + '-info'
+
+  const handleChange = (
+    e: React.ChangeEvent<HTMLInputElement> &
+      React.ChangeEvent<HTMLTextAreaElement>
+  ): void => {
+    const len = e.target.value.length
+    setState({
+      characterLength: len,
+      limitMessage: limitMessage(len),
+      invalidMessage: len > maxLength,
+    })
+  }
+
+  const handleBlur = (
+    e: React.FocusEvent<HTMLInputElement> &
+      React.FocusEvent<HTMLTextAreaElement>
+  ): void => {
+    const validationMessage = state.invalidMessage
+      ? 'The content is too long.'
+      : ''
+    e.target.setCustomValidity(validationMessage)
+  }
+
+  let inputComponent
+  if (isTextArea) {
+    const ref = textRef as TextareaRef
+    // How will we know if the label or hint has an id value should we add it to the describedby?
+    inputComponent = (
+      <Textarea
+        id={id}
+        data-testid="characterCountInput"
+        name={name}
+        className={classes}
+        defaultValue={defaultValue}
+        onChange={handleChange}
+        onBlur={handleBlur}
+        aria-describedby={messageId}
+        inputRef={ref}
+        {...inputProps}
+      />
+    )
+  } else {
+    const ref = textRef as TextInputRef
+    inputComponent = (
+      <TextInput
+        id={id}
+        data-testid="characterCountInput"
+        name={name}
+        type="text"
+        className={classes}
+        defaultValue={defaultValue}
+        onChange={handleChange}
+        onBlur={handleBlur}
+        aria-describedby={messageId}
+        inputRef={ref}
+        {...inputProps}
+      />
+    )
+  }
+
+  return (
+    <>
+      {/* Moves maxlength attribute off of input so it is not a hard entry limit */}
+      <div className="usa-character-count" data-maxlength={maxLength}>
```

Comment:
You don't need to include data-maxLength in the rendered component because this isn't an actual HTML attribute.

Reply:
I can remove it, was just mirroring their implementation.